docs/proposals/hugepages.md

```diff
@@ -0,0 +1,249 @@
+is accessed, 2M/4k = 512 TLB entries are needed to map the address range.  Each
+TLB miss results in an expensive walk of the page table.  However, if the
+allocation is backed by a 2M huge page by, only 1 TLB entry is required
+resulting in a highly likelihood that entry will remain in the cache and hit on
+accesses to the entire 2M structure.
+
+On x86_64, there are two huge page sizes: 2MB and 1G.  1G huge pages are also
+called gigantic pages.  1G must be enabled on kernel boot line with
+```hugepagesz=1G```. Huge pages, especially 1G ones, should to be allocated
+early before memory fragments (i.e. at/near boot time) to increase the
+likelihood that they can be allocated successfully with minimal memory migration
+(i.e. defreg) required.
+
+## Use Cases
+
+The class of applications that benefit from huge pages are typically have
+- A large memory working set
+- A sensitivity to memory access latency
+
+Example applications include:
+- Java applications can back the heap with huge pages using the ```-XX:+UseLargePages``` option.
+- In-memory databases
+
+Applications can generally use huge pages by calling
+- ```mmap()``` with ```MAP_ANONYMOUS | MAP_HUGETLB``` and use it as anonymous memory
+- ```shmget()``` with ```SHM_HUGETLB``` and use it as a shared memory segment (see Known Issues).
+
+### Pod Specification
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example
+spec:
+  containers:
+...
+    resources:
+      requests:
+	    hugepages: "10"
+      limits:
+	    hugepages: "10"
+  nodeSelector:
+    kubernetes.io/huge-page-size: "2M"
+```
+
+Huge pages can not be overcommitted on a node.
+
+While a system may support multiple huge pages sizes, it is assumed that nodes
+configured with huge pages will only use one huge page size, namely the default
+page size in ```cat /proc/meminfo | grep Hugepagesize```.  In Linux, this is 2M
+unless overridden by ```default_hugepagesz=1g``` in the kernel boot parameters.
+
+The huge page size for the node will be reported by the kubelet as a label
+```alpha.kubernetes.io/huge-page-size``` on the node resource.  This is done
+because there are a variety of huge page sizes across different hardware
+architecture and making a new resource field for each size doesn't scale.  Pods
+can do a nodeSelector on this label to land on a system with a particular huge
+page size.  This is similiar to how the ```beta.kubernetes.io/arch``` label
+operates.
+
+## Limits and Quota
+
+LimitRange should be able to define minimum and maximum constraints for huge
+pages, and Quota should be able to count them.
+
+## Implementation
+
+### Phase 0: Design Agreement
+
+**Target 1.5**
+
+Get design approval
+
+### Phase 1: Add huge page support
+
+**Target 1.5+**
+
+Implement huge page support with pod-level cgroups to enforce per-pod huge page
+limits (not yet available).  Enforcing huge page limits with pod-level cgroups
+avoids, at least temporarily, the need for 1) docker to support the
```

Comment:
actually, in this case, i do mean `docker` the command, not `Docker` the company.  i'll put code ticks around it.